crates/astria-conductor/src/driver.rs

```diff
@@ -128,7 +139,15 @@ impl Driver {
             NetworkEvent::Message(msg) => {
                 debug!("received gossip message: {:?}", msg);
                 let block = SequencerBlock::from_bytes(&msg.data)?;
-                // TODO: validate this block!!!!!
+
+                // validate block received from gossip network
+                // note: `validate_sequencer_block` is async, but `handle_network_event` cannot be
+                // async due to libp2p restrictions.
+                let handle = tokio::runtime::Handle::current();
+                handle
+                    .block_on(self.block_verifier.validate_sequencer_block(&block))
```

Comment:
This makes me very uneasy. Can you link some documentation that explains this a bit further (namely the bit about `handle_network_event` not being async)?


If this is the case then `Driver::run` or the part that is interacting with the p2p network should be put into a `tokio::task::spawn_blocking`

Reply:
yeah, so to handle this async I need to make `Driver::run()` async, which then causes this error inside where `Driver::run()` is spawned here:
```
    tokio::task::spawn(async move {
        if let Err(e) = driver.run().await {
            panic!("Driver error: {}", e)
        }
    });
```

```
future cannot be sent between threads safely
the trait `Sync` is not implemented for `(dyn libp2p_core::transport::boxed::Abstract<(libp2p_identity::peer_id::PeerId, libp2p_core::muxing::boxed::StreamMuxerBox)> + Send + Unpin + 'static)`rustc[Click for full compiler diagnostic](rust-analyzer-diagnostics-view:/diagnostic%20message%20%5B0%5D?0#file%3A%2F%2F%2Fhome%2Fe%2Fastria%2Fcrates%2Fastria-conductor%2Fsrc%2Fbin%2Fastria-conductor.rs)
driver.rs(120, 55): future is not `Send` as this value is used across an await
spawn.rs(163, 21): required by a bound in `tokio::spawn`
```

this makes me think the driver shouldn't contain an instance of `GossipNetwork` and they should communicate over a channel or something, what do you think?

Reply:
Yeah, I agree. Let's put the loop over the gossip stream into its own dedicated task, since this need not all be done in the same `select!` and it can just be its own thing. So it can get its clone of `executor_tx` etc

Inside the loop over the gossip stream we could make use of a tokio `LocalSet` to handle the events asynchronously. But this could be left to another PR.

I'll investigate what's with the trait bounds when I'm back from vacation

Reply:
sounds good, I opened an issue: #111

Reply:
fixed with sync_wrapper :DD